File: stacks/valid-parentheses/valid-parentheses.java
class Solution {
  public boolean isValid(String s) {

    Map<Character, Character> map = new HashMap<>();
    map.put(')', '(');
    map.put('}', '{');
    map.put(']', '[');

    Stack<Character> stack = new Stack<>();

    for (char c : s.toCharArray()) {
      if (!stack.isEmpty() && stack.peek() == map.get(c)) {
        stack.pop();
      } else {
        stack.push(c);
      }
    }

    return stack.isEmpty();
  }
}

File: stacks/valid-parentheses/valid-parentheses.ts
function isValid(s: string): boolean {
  const map: Record<string, string> = {
    ')': '(',
    '}': '{',
    ']': '[',
  };

  const stack: string[] = [];

  for (let c of s) {
    if (stack.length && stack[stack.length - 1] === map[c]) {
      stack.pop();
    } else {
      stack.push(c);
    }
  }

  return stack.length === 0;
}
